service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,209 @@
+package matching
+
+import (
+	"fmt"
+	"slices"
+	"time"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common"
+	hlc "go.temporal.io/server/common/clock/hybrid_logical_clock"
+	"go.temporal.io/server/common/util"
+)
+
+// checkAssignmentConditions returns an error if the new set of assignment rules don't meet the following requirements:
+// - No more rules than dynamicconfig.VersionAssignmentRuleLimitPerQueue
+// - No assignment rule's TargetBuildId can be a member of an existing version set
+// - If there existed an "unfiltered" assigment rule (which can accept any build id), at least one must still exist
+// - To override the unfiltered assignment rule requirement, the user can specify force = true
+func checkAssignmentConditions(g *persistencepb.VersioningData, maxARs int, force, hadUnfiltered bool) error {
+	rules := slices.Clone(g.GetAssignmentRules())
+	if maxARs > 0 && countActiveAR(rules) > maxARs {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update exceeds number of assignment rules permitted in namespace dynamic config (%v/%v)", len(rules), maxARs))
```

Comment:
```suggestion
		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update exceeds number of assignment rules permitted in namespace (%v/%v)", len(rules), maxARs))
```
Since these errors are likely shown to the user, let's try not include to much of server-specific details.